### PR TITLE
Experimental parameter importance feature.

### PR DESCRIPTION
**Pull request by hvy**
An experimental parameter importance feature depending on [automl/fANOVA](https://github.com/automl/fanova).

A concern is that the above fANOVA depends on [`pyrfr`](https://github.com/automl/random_forest_run) which requires SWIG for its bindings, but this PR could start a discussion whether this feature is useful at all and if so what API we should provide.

#### Note

The following are excluded from this PR and are meant to be addressed in later PRs. The interfaces introduced in this PR are therefore experimental.

- Standard deviations of importances.
- Pairwise importances.
- Parameters for configuring evaluators.

#### Experiments

<details>
  <summary>LightGBM, 100 trials, `RandomSampler`</summary>
  
  
</details>
<details>
  <summary>LightGBM, 100 trials, `TpeSampler`</summary>
  
  
</details>
<details>
  <summary>PyTorch, 100 trials, `RandomSampler`</summary>
  
  
</details>
<details>
  <summary>PyTorch, 100 trials, `TpeSampler`</summary>
  
  
</details>
<details>
  <summary>ffmpeg, 150 trials, `RandomSampler`</summary>
  
  
</details>
<details>
  <summary>ffmpeg, 150 trials, `TpeSampler`</summary>
  
  
</details>

`PermutationImportanceEvaluator` and `RandomForestFeatureImportanceEvaluator` were experimental evaluators besides fANOVA based on scikit-learn's ([`permutation_importance`](https://scikit-learn.org/stable/modules/generated/sklearn.inspection.permutation_importance.html) and [`RandomForestRegressor.feature_importances_`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor.feature_importances_)), now removed from this PR but can be recognized from the commit history. Categorical importances were encoded as one-hot sparse matrices which was suboptimal. fANOVA's random forest properly handles/splits on categorical features so that is not an issue.

##### LightGBM Ablation Study

Below result shows that fANOVA when applied to a study of 100 trials with the `RandomSampler`, seemingly manages to assess the importances for most hyperparameters. Note that although `feature_fraction` is ranked high by fANOVA but still leads to high accuracy when omitted, the variance is quite large.

Configurations are as follows.

- Trials: 100
- Sampler: `RandomSampler`
- k (repetitions per parameter): 30

For each hyperparameter, run k studies with that hyperparameter being fixed, optimizing all other hyperparameters. The parameter value for the fixed hyperparameter is choosen randomly from a uniform distribution. The total number of trials is thus 100 * 30 * 7 (number of hyperparameters).
Error bars are standard deviations among k studies.

**LightGBM, 100 trials, `RandomSampler`**

ablation_results

Order by importance assessed by the `FanovaImportanceEvaluator`. Top row is most important.

#### TODO

- [x] Experiments ~to make sure it's stable and that `fANOVA` does not yield NaNs or errors~
- [x] Fix interface
- [x] Write more tests
- [x] Reconsider third party dependencies
- [x] Document
- [x] Paste benchmark results here.

### CODE REVIEW

**Comment by hvy** [comment]
Leaving a note. Latest swig 4.* yields a segmentation fault when instantiating the pyrfr model in `fANOVA.__init__`. Downgrading to swig 3.* solves the issue.

File: optuna/importance/_fanova.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    ConfigurationSpace = None
+    Hyperparameter = None
+    fANOVA = None
+
+    _import_error = e
+    _available = False
+
+
+class FanovaImportanceEvaluator(BaseImportanceEvaluator):
+    def __init__(self) -> None:
```
Comment (hvy) [inline optuna/importance/_fanova.py]: Users probably want to specify arguments passed to `fANOVA.__init__` such as the random seed `seed`. Should we support all arguments in this PR?

File: optuna/importance/__init__.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from typing import Optional
+
+from optuna._experimental import experimental
+from optuna.importance._base import BaseImportanceEvaluator
+from optuna.importance._fanova import FanovaImportanceEvaluator
+from optuna.study import Study
+
+
+@experimental("1.3.0")
+def get_param_importances(
```
Comment (hvy) [inline optuna/importance/__init__.py]: Alternatives that might be more informative / better in some other sense

- `evaluate_param_importances`
- `evaluate_importances`
- `evaluate`

Other suggestions are highly welcomed.

File: optuna/importance/_fanova.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    def evaluate(self, study: Study, params: Optional[List[str]]) -> Dict[str, float]:
+        distributions = get_distributions(study, params)
+        params_data, values_data = get_study_data(study, distributions)
+
+        evaluator = fANOVA(
+            X=params_data,
+            Y=values_data,
+            config_space=_get_configuration_space(distributions),
+            max_features=max(1, int(params_data.shape[1] * 0.7)),
```
Comment (sile) [inline optuna/importance/_fanova.py]: I'm just curious why you adopted this expression instead of using the default value of `max_features`.

File: optuna/samplers/__init__.py [hunk 4]
```diff
@@ -21,6 +23,8 @@ def intersection_search_space(study):
     neither is included in the resulting search space
     (i.e., the parameters with dynamic value ranges are excluded).
 
+    TODO(hvy): Document arguments.
```
Comment (sile) [inline optuna/samplers/__init__.py]: Looks like an important TODO.

File: optuna/importance/_base.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    return params, values
+
+
+def _check_evaluate_args(study: Study, params: Optional[List[str]]) -> None:
+    completed_trials = list(filter(lambda t: t.state == TrialState.COMPLETE, study.trials))
+    if len(completed_trials) == 0:
+        raise ValueError("Cannot evaluate parameter importances without completed trials.")
+
+    if params is not None:
+        if not isinstance(params, (list, tuple)):
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/importance/_base.py]: We don't have to, but I think it's quite common among Python users to expect a simple enumeration (container of strings) to also accept a tuple. Do you think it's better to be strict?

File: optuna/importance/_fanova.py (hunk 3 above)
Comment (hvy) [reply optuna/importance/_fanova.py]: It's to handle cases where `X` only has a single or no column (either or both, sorry I can't recall but could confirm). An alternative would be to check the shape of `X` and do an early return or similar.

File: optuna/importance/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/importance/_base.py]: Thanks. That makes sense. It seems okay to keep this code as it is.

[the remark this answers is not in the document]
Comment (sile) [reply optuna/importance/_base.py]: **FYI** I found that `Trial.suggest_categorical` uses `Sequence[...]` to allow lists and tuples (and other sequence types).

https://github.com/optuna/optuna/blob/b72426ffd138d1d4770077f6e45186c15a37b900/optuna/trial.py#L47-L48

File: optuna/importance/_fanova.py (hunk 3 above)
Comment (sile) [reply optuna/importance/_fanova.py]: I see. Thanks!

**Comment by g-votte** [comment]
@hvy 
I have encountered a `RuntimeError` inside `fanova` with the following test code. Although I'm not 100% sure about the cause, `fanova` seems not to be compatible with just one trial, and we may want to handle this edge case.

```python
def test_one_trial() -> None:
    def objective(trial: Trial) -> float:
        x1 = trial.suggest_uniform("x1", 1, 3)
        return x1 ** 2

    study = create_study()
    study.optimize(objective, n_trials=1)

    get_param_importances(study, evaluator=FanovaImportanceEvaluator(),
                          params=["x1"])
```

File: docs/source/reference/importance.rst [hunk 6]
```diff
@@ -0,0 +1,12 @@
+.. module:: optuna.importance
+
+Hyperparameter Importance
+=========================
+
+.. autofunction:: get_param_importances
+
+.. autoclass:: BaseImportanceEvaluator
```
[the remark this answers is not in the document]
Comment (sile) [reply docs/source/reference/importance.rst]: **FYI** [BaseSampler](https://optuna.readthedocs.io/en/stable/reference/samplers.html#optuna.samplers.BaseSampler) is exposed in the doc.

[the remark this answers is not in the document]
Comment (g-votte) [reply docs/source/reference/importance.rst]: @sile 
Thanks for your comment. Isn't it because `BaseSampler` is supposed to be used by a library user?
Either is actually fine for me, but I just think we should be consistent.

[the remark this answers is not in the document]
Comment (sile) [reply docs/source/reference/importance.rst]: I basically agree with @g-votte option. Because, unlike `BaseSampler`, it's not expected that users wll implement their own custom importance evaluators, I think that it's not needed to expose `BaseImportanceEvaluator ` to users at the time being.

File: optuna/importance/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/importance/_base.py]: I see, thanks for the note. `Sequence` contains 'str' types as well so I'm not sure it's appropriate here nor in `Trial.suggest_categorical` 🤔 .

[the remark this answers is not in the document]
Comment (sile) [reply optuna/importance/_base.py]: Hmm, `suggest_categorical` uses `Sequence[CategoricalChoiceType]`, a specialized type of `Sequence`,  as the type of `choices` argument. I'm not sure but it seems differ from `str`(maybe this is another specialized type of `Sequence`?).

For instance, I applied `mypy` to the following code:
```python
from typing import Sequence

def foo(seq: Sequence[int]) -> None:
    pass

foo("hello")
```

Then, `mypy` emitted the following error:
```
foo.py:8: error: Argument 1 to "foo" has incompatible type "str"; expected "Sequence[int]"
Found 1 error in 1 file (checked 1 source file)
```

So, it seems `mypy` recognizes as `Sequence[int]` and `str` are different types.

[the remark this answers is not in the document]
Comment (sile) [reply optuna/importance/_base.py]: Note that I wrote the above comment just to clarity the behavior of type hint and `mypy`. I think that it's okay to keep using `List[str]` as the type of `param` argument because list type may be more familiar with normal users and sufficient for typical use cases.

File: optuna/importance/_fanova.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+        `An Efficient Approach for Assessing Hyperparameter Importance
+        <http://proceedings.mlr.press/v32/hutter14.html>`_.
+    """
+
+    def __init__(self) -> None:
+        _check_fanova_availability()
+
+    def evaluate(self, study: Study, params: Optional[List[str]]) -> Dict[str, float]:
+        distributions = get_distributions(study, params)
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/importance/_fanova.py]: I thought of that as well. However, I was ambivalent where to put these possibly overlapping logic. It could be placed in `get_param_importance`, but also `BaseImportanceEvaluator`. For now, until we have more evaluators, I'd like to keep the `get_param_importance` wrapper as thin as possible to not scatter the logic of the importance evaluation. I left a TODO comment. Does this seem acceptable?

File: docs/source/reference/importance.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (hvy) [reply docs/source/reference/importance.rst]: Fixed!

File: optuna/importance/_base.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            )
+
+    assert distributions is not None  # Required to pass mypy.
+    distributions = OrderedDict(
+        sorted(distributions.items(), key=lambda name_and_distribution: name_and_distribution[0])
+    )
+    return distributions
+
+
+def get_study_data(
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/importance/_base.py]: Fixed!

**Comment by hvy** [comment]
@g-votte thanks for the catch about the single trial edge case. It is now checked before the call to fANOVA as implemented here https://github.com/optuna/optuna/pull/946/commits/e1a9ef0e571bcbc7de859335d57d54d649967418.

**Comment by hvy** [comment]
PTAL.

File: optuna/importance/_fanova.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/importance/_fanova.py]: Thanks for letting me know your thought. It's acceptable.

File: optuna/importance/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/importance/_base.py]: Ah, and if you don't pass any arguments to the type hint of `Sequence`, `mypy` don't complain when passing `str`s. That makes sense. Thank you for the clarification. However, I agree that a simple `List` is more user friendly.

File: optuna/samplers/__init__.py (hunk 4 above)
Comment (hvy) [reply optuna/samplers/__init__.py]: Addressed here https://github.com/optuna/optuna/pull/1053.
